molr-mole-core/src/main/java/io/molr/mole/core/support/OngoingMissionRun.java

```diff
@@ -41,4 +44,107 @@ public OngoingMissionRun and() {
                 .map(MissionState::result);
     }
 
+    public Result whenFinished(Duration timeout) {
+        return awaitFinished().block(timeout);
+    }
+
+    public Result whenFinished() {
+        return awaitFinished().block();
+    }
+
+    public Mono<MissionHandle> asyncHandle() {
+        return handle;
+    }
+
+    public MissionHandle awaitHandle() {
+        return handle.block();
+    }
+
+    public void instruct(StrandCommand command) {
+        handle.subscribe(h -> mole.instructRoot(h, command));
+    }
+
+    public void await(RunStateValidator validator) {
+        filterState(validator, handle, mole)
+                .blockFirst();
+    }
+
+    public void await(RunStateValidator validator, Duration timeout) {
+        filterState(validator, handle, mole)
+                .blockFirst(timeout);
+    }
+
+    private static Flux<MissionState> filterState(RunStateValidator validator, Mono<MissionHandle> handle, Mole mole) {
+        return handle
+                .flatMapMany(mole::statesFor)
+                .filter(s -> validator.test(s.runState()));
+    }
+
+    public ReturnHelper returnResult() {
+        return new ReturnHelper<MissionState, Result>(Result.class);
+    }
+
+    public ReturnHelper returnState() {
+        return new ReturnHelper<MissionState, MissionState>(MissionState.class);
+    }
+
+    public static RunStateValidator runState(RunState runState) {
+        return RunStateValidator.validateState(runState);
+    }
+
+    public class ReturnHelper<T, R> {
+        private Function<T, R> function;
+
+        public ReturnHelper(Class<R> type) {
```

Comment:
Couldnt you just pass in the function instead of the type? This would save this if ...

Reply:
Yes, I too think there's some scope for refactoring in this inner class. Will think more about it and update it. ;-)